### PR TITLE
Evaluate Class methods calls at compile time when possible

### PR DESCRIPTION
**Pull request by nbhuiyan**
Several java/lang/Class methods, such as Class.isAnonymous, Class.hashCode(), etc. can often be determined at compile time, and is safe to do so if the class is a constant class that is known at compile time. This changeset introduces the following:
 * Recognize various Class methods to be folded
 * VP transformation that checks if the call can be folded, and obtains the values to replace the call node

Signed-off-by: Nazim Bhuiyan <nubhuiyan@ibm.com>

### CODE REVIEW

**Comment by vijaysun-omr** [comment]
Thanks for this work, did you consider the `getAnno*` family of methods in the `Class` class for folding ?

File: runtime/compiler/env/VMJ9.cpp [hunk 1]
```diff
@@ -6768,20 +6768,22 @@ TR_J9VMBase::isInterfaceClass(TR_OpaqueClassBlock * clazzPointer)
...
-   if (isClassArray(clazzPointer))
-      return false;
-   bool success = javaLangClassGetModifiersImpl(clazzPointer, modifiersForClass);
-   if (!success)
-      return false;
-   bool isModFlagSet = (modifiersForClass & J9AccEnum) ? true : false;
-   TR_OpaqueClassBlock *enumClass = getClassFromSignature("java/lang/Enum", 14, method);
-   TR_OpaqueClassBlock *superClass = getSuperClass(clazzPointer);
-   return (isModFlagSet && (enumClass == superClass));
+   return (TR::Compiler->cls.romClassOf(clazzPointer)->modifiers & J9AccEnum) ? true : false;
```
Comment (vijaysun-omr) [inline runtime/compiler/env/VMJ9.cpp:6773]: Can you please add a comment here clarifying why this change in logic was made ?

File: runtime/compiler/optimizer/J9ValuePropagation.cpp [hunk 2]
```diff
@@ -1381,6 +1381,92 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
...
+            if (rm == TR::java_lang_Class_isHidden)
+               result = comp()->fej9()->isHiddenClass(jlClassChildConstraint->getClass());
+            else if (rm == TR::java_lang_Class_isAnonymousClass)
+               result = comp()->fej9()->isAnonymousClass(jlClassChildConstraint->getClass());
+            else if (rm == TR::java_lang_Class_isEnum)
+               result = comp()->fej9()->isEnumClass(jlClassChildConstraint->getClass());
+            else if (rm == TR::java_lang_Class_isSynthetic)
+               result = comp()->fej9()->isSyntheticClass(jlClassChildConstraint->getClass());
+            else if (rm == TR::java_lang_Class_isArray)
+               result = comp()->fej9()->isPrimitiveArray(jlClassChildConstraint->getClass());
```
Comment (vijaysun-omr) [inline runtime/compiler/optimizer/J9ValuePropagation.cpp:1411]: Is this the right semantic if folding is done ? i.e. would we want to fold to a constant `false` if we had a reference array in this case ?

**Comment by vijaysun-omr** [comment]
Do you feel we should encourage inlining of callees when a constant class or string is being passed in ? I believe there is already a heuristic to encourage for numerical constants (and possibly even string constants) but you are obviously folding more now.

File: runtime/compiler/optimizer/J9ValuePropagation.cpp (hunk 2 above)
Comment (nbhuiyan) [reply runtime/compiler/optimizer/J9ValuePropagation.cpp:1411]: Oh, I missed the second part of the check here, which would be `|| comp()->fej9()->isReferenceArray(jlClassChildConstraint->getClass());`. I will fix that.

**Comment by nbhuiyan** [comment]
@vijaysun-omr 

> did you consider the getAnno* family of methods in the Class class for folding 

Yes, these are some the annotation related methods I considered:
* `isAnnotation` - this could possibly be implemented via a VM helper similar to the others in this PR, by checking if `J9AccAnnotation` bit in ROMClass->modifiers field is set.
* `getAnnotations` - this is more complicated to implement as it involves a few steps to obtain the annotation array. This is still something I am trying to implement and will be part of a different PR focused on the annotation-related queries. See Java implementation below:

```
    public Annotation[] getAnnotations() {
        return AnnotationParser.toArray(annotationData().annotations);
    }
```
* `getAnnotation` 
* `isAnnotationPresent` 

`getAnnotation` and `isAnnotationPresent` would depend on how `getAnnotations` is implemented, and should be fairly straightforward once I have a way to obtain the annotations map.

File: runtime/compiler/env/VMJ9.cpp (hunk 1 above)
Comment (nbhuiyan) [reply runtime/compiler/env/VMJ9.cpp:6773]: I will be reverting this. The previous implementation also checked that the class extends `java/lang/Enum`, without which `isEnum` should return false regardless of whether the `ENUM` flag is set or not.

**Comment by nbhuiyan** [comment]
> Do you feel we should encourage inlining of callees when a constant class or string is being passed in ? I believe there is already a heuristic to encourage for numerical constants (and possibly even string constants) but you are obviously folding more now.

Yes, I think that is a good idea to try out. Could result in more optimized code that would have otherwise not been possible if the callee with the constant class/string  is not inlined. From a brief search, there is `TR_MultipleCallTargetInliner::applyArgumentHeuristics` implementing special handling for constant classes in the args.

**Comment by nbhuiyan** [comment]
@vijaysun-omr Requesting another round of review.

For encouraging inlining of callees with constant classes in the args, the mechanism in `TR_MultipleCallTargetInliner::applyArgumentHeuristics` gets used already. For callees containing constant classes, `applyArgumentHeuristics` sets the call site weight to 10, and then subtracts numArgs*4 from it, often resulting in negative callsite weights in my experiments. I also experimented with a similar mechanism for constant strings in callee args that I will include as part of a future PR implementing compile time folding of some of the methods of `java/lang/String`.

**Comment by vijaysun-omr** [comment]
Jenkins test sanity all jdk8,jdk11,jdk19

File: runtime/compiler/optimizer/InlinerTempForJ9.cpp [hunk 3]
```diff
@@ -4780,6 +4780,18 @@ TR_J9InlinerPolicy::supressInliningRecognizedInitialCallee(TR_CallSite* callsite
          return true;
       case TR::java_lang_Class_cast:
          return true; // Call will be transformed into checkcast
+      case TR::java_lang_Object_hashCode:
+         if (callNode &&
```
Comment (vijaysun-omr) [inline runtime/compiler/optimizer/InlinerTempForJ9.cpp:4792]: Is it assumed that this call node will have a direct call opcode ? Otherwise the first child will be a load of the vft, not the actual first argument.

It is probably better to check `getNumChildren() > 0` in order to determine that one can access the first child. It is possible in theory that some transformation sets the number of children to N and does not null out the child pointers for any value above N.

File: runtime/compiler/optimizer/InlinerTempForJ9.cpp [hunk 4]
```diff
@@ -4780,16 +4780,23 @@ TR_J9InlinerPolicy::supressInliningRecognizedInitialCallee(TR_CallSite* callsite
...
+      case TR::java_lang_Class_isSynthetic:
+      case TR::java_lang_Class_isArray:
+      case TR::java_lang_Class_isPrimitive:
+      case TR::java_lang_Class_isAnnotation:
       case TR::java_lang_Object_hashCode:
-         if (callNode &&
-             callNode->getFirstChild() &&
-             callNode->getFirstChild()->getOpCode().hasSymbolReference())
+         if (callNode
+             && callNode->getNumChildren() > 0)
```
Comment (vijaysun-omr) [inline runtime/compiler/optimizer/InlinerTempForJ9.cpp:4792]: Maybe more has been deleted here than was intended ? I think it is okay to check that the number of children is > 0 but did you also intend to avoid checking if the first (or maybe it is last since that was changed in the last commit I think ?) child has a symbol reference before accessing it ?

Comment (nbhuiyan) [reply runtime/compiler/optimizer/InlinerTempForJ9.cpp]: I re-added checking if the last child has symref before accessing it, now that I realize that `getSymbolReference` would fail an assertion instead of just returning NULL.

Comment (nbhuiyan) [reply runtime/compiler/optimizer/InlinerTempForJ9.cpp]: I have also removed `java/lang/Class.isInstance` from here since the decision to inline that is done during ILGen (through native call inlining). I will contribute that check to skip inlining that as well as the VP transformation in a separate PR.

**Comment by vijaysun-omr** [comment]
Jenkins test sanity all jdk8,jdk11,jdk19

File: runtime/compiler/optimizer/InlinerTempForJ9.cpp [hunk 5]
```diff
@@ -4780,6 +4780,18 @@ TR_J9InlinerPolicy::supressInliningRecognizedInitialCallee(TR_CallSite* callsite
          return true;
       case TR::java_lang_Class_cast:
          return true; // Call will be transformed into checkcast
+      case TR::java_lang_Object_hashCode:
+         if (callNode &&
+             callNode->getFirstChild() &&
+             callNode->getFirstChild()->getOpCode().hasSymbolReference())
+            {
+            TR::SymbolReference * classChildSymRef = callNode->getFirstChild()->getSymbolReference();
```
Comment (vijaysun-omr) [inline runtime/compiler/optimizer/InlinerTempForJ9.cpp]: I had a prior comment about whether this code assumes the call will be a direct one. If so, why is it okay to assume this ? I ask, because the code in the other commit gets the last child instead of the first one and that would skip over the vft load that would be the first child of an indirect call.

Comment (nbhuiyan) [reply runtime/compiler/optimizer/InlinerTempForJ9.cpp]: Yes, this code was assuming the call will be a direct call, and this is based on the generated trees for all the methods except Object.hashCode. Object.hashCode gets transformed by VP from an indirect call to a direct call prior to the new hashcode folding transformation working on it. We also get the direct call version of Object.hashcode if the call node gets added in the root method from an inlined callee. However, at this stage of the compilation, since Object.hashCode call can be an indirect call, it was incorrect to use the first child here since it can be a vft load child. Using the last child is the correct approach to deal with that.

**Comment by pshipton** [comment]
Reverting due to https://github.com/eclipse-openj9/openj9/issues/16802
